Automattic/jetpack
# Social: Fix parallel social connections requests mixing up UI state

Replaces #38367

> On slower connections there is a scenario which could happen that messes up the connection state:
> - On page load the `test_connections` call fires, but takes 10-15 seconds to complete
> - In the meantime the user adds/removes a connection, changing the state
> - If the addition/remove of connection finishes faster, the `test_connections` call will mess up the state when it finishes, it will either remove the new connection from the UI, or add back the removed one.
> 

## Proposed changes:

* Use [AbortController](https://developer.mozilla.org/en-US/docs/Web/API/AbortController) in [`wp.apiFetch`](https://github.com/WordPress/gutenberg/blob/83f671f059531da6945e46c30f1870e1a81a2df5/packages/api-fetch/README.md#aborting-a-request) to cancel the existing test connections request if a connection creation/updation/deletion request is fired. The test connection request is fired again after the request finishes.
* Make the test connections request wait if a connection creation/updation/deletion request is already in progress

### Other information:


## Jetpack product discussion



## Does this pull request change what data or activity we track or use?



## Testing instructions:






* Goto post editor
### Case 1: Connections test request is not fired until an in-flight connection creation/updation/deletion request is in progress 


* To simulate a slow connection creation request, add `sleep( 10 );` on [line 591 in `packages/publicize/src/class-rest-controller.php`](https://github.com/Automattic/jetpack/blob/ab29101d1cb3f88b9b9e66a22f3ca56e74e02db4/projects/packages/publicize/src/class-rest-controller.php#L591)
* Open the connections management modal
* Connect an account and wait on the confirmation screen
* Open the Network panel in the browser dev tools and clear all the previous requests
* Confirm the connection and quickly switch to a different tab and then switch back to the editor tab
* Confirm that the test connections request is not fired until the connection creation request finishes
* Confirm the same with updating/globalizing and deleting a connection requests

### Case 12: Connections test request is cancelled when a new connection creation/updation/deletion request is fired. 

* To simulate a slow connections test request, add this on [line 272 in `packages/publicize/src/class-publicize.php`](https://github.com/Automattic/jetpack/blob/ab29101d1cb3f88b9b9e66a22f3ca56e74e02db4/projects/packages/publicize/src/class-publicize.php#L272)
    ```php
    if ( ! empty( $args['test_connections'] ) ) {
		sleep( 10 );
	}
    ```
* Now add a connection again and wait on the confirmation screen
* Switch to a different tab and then switch back to the editor tab
* See the connection test request getting fired
* Quickly click on Confirm to add the connection
* Confirm that the in-flight test connection request gets canceled
* Wait for the connection creation request to finish
* Confirm that the test connection request now gets fired after the connection creation request finishes


https://github.com/user-attachments/assets/bd3608e5-9139-49ef-8974-6890048106ed
